File: cpp/src/evolution.cpp
// ============================================================================
// Evolution -- neuro-symbolic graph learning (feature-gated).
// ============================================================================

#ifdef BRAIN_EVOLUTION

#include "brain/evolution.hpp"
#include <sqlite3.h>
#include <nlohmann/json.hpp>
#include <algorithm>
#include <cmath>
#include <sstream>

using json = nlohmann::json;

namespace brain {

// Helper: clamp a float between lo and hi
static inline float clampf(float v, float lo, float hi) {
    return std::max(lo, std::min(hi, v));
}

// ---- EvolutionState ----

EvolutionState::EvolutionState() {}

void EvolutionState::record_feedback(FeedbackSignal signal) {
    feedback_buffer.push_back(std::move(signal));
}

void EvolutionState::train_step() {
    for (const auto& signal : feedback_buffer) {
        float delta = signal.useful ? learning_rate : -learning_rate;

        for (int64_t id : signal.memory_ids) {
            auto it = node_weights.find(id);
            if (it == node_weights.end()) {
                node_weights[id] = clampf(1.0f + delta, 0.1f, 2.0f);
            } else {
                it->second = clampf(it->second + delta, 0.1f, 2.0f);
            }
        }

        for (const auto& pair : signal.edge_pairs) {
            auto it = edge_weights.find(pair);
            if (it == edge_weights.end()) {
                edge_weights[pair] = clampf(1.0f + delta, 0.1f, 2.0f);
            } else {
                it->second = clampf(it->second + delta, 0.1f, 2.0f);
            }
        }
    }

    feedback_buffer.clear();
    ++generation;
}

float EvolutionState::get_node_weight(int64_t id) const {
    auto it = node_weights.find(id);
    return it == node_weights.end() ? 1.0f : it->second;
}

float EvolutionState::get_edge_weight(int64_t source, int64_t target) const {
    auto it = edge_weights.find({source, target});
    return it == edge_weights.end() ? 1.0f : it->second;
}

std::string EvolutionState::save_state(sqlite3* db) const {
    if (!db) return "db is null";

    // Build JSON: keys as strings (node weights as id string, edge weights as "src,tgt")
    json j;
    j["generation"] = generation;
    j["learning_rate"] = learning_rate;

    json nw = json::object();
    for (const auto& [id, w] : node_weights) {
        nw[std::to_string(id)] = w;
    }
    j["node_weights"] = nw;

    json ew = json::object();
    for (const auto& [pair, w] : edge_weights) {
        std::string key = std::to_string(pair.first) + "," + std::to_string(pair.second);
        ew[key] = w;
    }
    j["edge_weights"] = ew;

    std::string json_str = j.dump();
    const void* blob = json_str.data();
    int blob_size = static_cast<int>(json_str.size());

    sqlite3_stmt* stmt = nullptr;
    int rc = sqlite3_prepare_v2(db,
        "INSERT OR REPLACE INTO brain_meta (key, value, updated_at) VALUES ('evolution_state', ?, datetime('now'))",
        -1, &stmt, nullptr);
    if (rc != SQLITE_OK) {
        return std::string("prepare failed: ") + sqlite3_errmsg(db);
    }

    sqlite3_bind_blob(stmt, 1, blob, blob_size, SQLITE_STATIC);
    rc = sqlite3_step(stmt);
    sqlite3_finalize(stmt);

    if (rc != SQLITE_DONE) {
        return std::string("step failed: ") + sqlite3_errmsg(db);
    }

    return "";
}

EvolutionState EvolutionState::load_state(sqlite3* db) {
    EvolutionState state;
    if (!db) return state;

    sqlite3_stmt* stmt = nullptr;
    int rc = sqlite3_prepare_v2(db,
        "SELECT value FROM brain_meta WHERE key = 'evolution_state'",
        -1, &stmt, nullptr);
    if (rc != SQLITE_OK) return state;

    if (sqlite3_step(stmt) != SQLITE_ROW) {
        sqlite3_finalize(stmt);
        return state;
    }

    const void* blob = sqlite3_column_blob(stmt, 0);
    int blob_size = sqlite3_column_bytes(stmt, 0);

    if (!blob || blob_size <= 0) {
        sqlite3_finalize(stmt);
        return state;
    }

    std::string json_str(static_cast<const char*>(blob), blob_size);
    sqlite3_finalize(stmt);

    try {
        json j = json::parse(json_str);

        state.generation = j.value("generation", 0u);
        state.learning_rate = j.value("learning_rate", 0.01f);

        if (j.contains("node_weights") && j["node_weights"].is_object()) {
            for (auto& [key, val] : j["node_weights"].items()) {
                try {
                    int64_t id = std::stoll(key);
                    state.node_weights[id] = val.get<float>();
                } catch (...) {}
            }
        }

        if (j.contains("edge_weights") && j["edge_weights"].is_object()) {
            for (auto& [key, val] : j["edge_weights"].items()) {
                auto comma = key.find(',');
                if (comma == std::string::npos) continue;
                try {
                    int64_t src = std::stoll(key.substr(0, comma));
                    int64_t tgt = std::stoll(key.substr(comma + 1));
                    state.edge_weights[{src, tgt}] = val.get<float>();
                } catch (...) {}
            }
        }
    } catch (...) {
        // Return fresh state on any parse error
        return EvolutionState{};
    }

    return state;
}

EvolutionStats EvolutionState::stats() const {
    return EvolutionStats{
        generation,
        node_weights.size(),
        edge_weights.size(),
        learning_rate
    };
}

} // namespace brain

// ============================================================================
// C++ Evolution Tests (compiled only when BRAIN_EVOLUTION is defined and
// the test binary is requested)
// ============================================================================

#ifdef BRAIN_EVOLUTION_TESTS

#include <cassert>
#include <cstdio>
#include <cmath>

// Minimal in-memory SQLite helper for tests
static sqlite3* open_test_db() {
    sqlite3* db = nullptr;
    sqlite3_open(":memory:", &db);
    sqlite3_exec(db,
        "CREATE TABLE brain_meta (key TEXT PRIMARY KEY, value BLOB, updated_at TEXT);",
        nullptr, nullptr, nullptr);
    return db;
}

static void test_default_weights() {
    brain::EvolutionState state;
    assert(state.get_node_weight(42) == 1.0f);
    assert(state.get_edge_weight(1, 2) == 1.0f);
    assert(state.learning_rate == 0.01f);
    assert(state.generation == 0);
    printf("[PASS] test_default_weights\n");
}

static void test_positive_feedback() {
    brain::EvolutionState state;
    brain::FeedbackSignal sig;
    sig.memory_ids = {1, 2};
    sig.edge_pairs = {{1, 2}};
    sig.useful = true;
    state.record_feedback(sig);
    state.train_step();

    float expected = 1.0f + 0.01f;
    assert(std::fabs(state.get_node_weight(1) - expected) < 1e-5f);
    assert(std::fabs(state.get_node_weight(2) - expected) < 1e-5f);
    assert(std::fabs(state.get_edge_weight(1, 2) - expected) < 1e-5f);
    assert(state.generation == 1);
    printf("[PASS] test_positive_feedback\n");
}

static void test_negative_feedback() {
    brain::EvolutionState state;
    brain::FeedbackSignal sig;
    sig.memory_ids = {5};
    sig.edge_pairs = {{5, 6}};
    sig.useful = false;
    state.record_feedback(sig);
    state.train_step();

    float expected = 1.0f - 0.01f;
    assert(std::fabs(state.get_node_weight(5) - expected) < 1e-5f);
    assert(std::fabs(state.get_edge_weight(5, 6) - expected) < 1e-5f);
    printf("[PASS] test_negative_feedback\n");
}

static void test_weight_bounds() {
    // Ceiling test
    brain::EvolutionState state;
    for (int i = 0; i < 200; ++i) {
        brain::FeedbackSignal sig;
        sig.memory_ids = {99};
        sig.edge_pairs = {{99, 100}};
        sig.useful = true;
        state.record_feedback(sig);
        state.train_step();
    }
    assert(state.get_node_weight(99) <= 2.0f);
    assert(std::fabs(state.get_node_weight(99) - 2.0f) < 1e-3f);
    assert(state.get_edge_weight(99, 100) <= 2.0f);

    // Floor test
    brain::EvolutionState state2;
    for (int i = 0; i < 200; ++i) {
        brain::FeedbackSignal sig;
        sig.memory_ids = {77};
        sig.edge_pairs = {{77, 78}};
        sig.useful = false;
        state2.record_feedback(sig);
        state2.train_step();
    }
    assert(state2.get_node_weight(77) >= 0.1f);
    assert(std::fabs(state2.get_node_weight(77) - 0.1f) < 1e-3f);
    printf("[PASS] test_weight_bounds\n");
}

static void test_save_load_roundtrip() {
    sqlite3* db = open_test_db();

    brain::EvolutionState state;
    state.node_weights[10] = 1.5f;
    state.node_weights[20] = 0.7f;
    state.edge_weights[{10, 20}] = 1.3f;
    state.generation = 7;

    std::string err = state.save_state(db);
    assert(err.empty());

    brain::EvolutionState loaded = brain::EvolutionState::load_state(db);
    assert(loaded.generation == 7);
    assert(std::fabs(loaded.get_node_weight(10) - 1.5f) < 1e-5f);
    assert(std::fabs(loaded.get_node_weight(20) - 0.7f) < 1e-5f);
    assert(std::fabs(loaded.get_edge_weight(10, 20) - 1.3f) < 1e-5f);
    assert(loaded.get_node_weight(999) == 1.0f);

    sqlite3_close(db);
    printf("[PASS] test_save_load_roundtrip\n");
}

int main() {
    test_default_weights();
    test_positive_feedback();
    test_negative_feedback();
    test_weight_bounds();
    test_save_load_roundtrip();
    printf("All evolution tests passed.\n");
    return 0;
}

#endif // BRAIN_EVOLUTION_TESTS
#endif // BRAIN_EVOLUTION

File: eidolon-lib/src/evolution.rs
// ============================================================================
// Evolution - neuro-symbolic graph learning (feature-gated).
// Per-node and per-edge learned weights trained from feedback signals.
//
// Feature gate: compile with --features evolution to enable.
// When disabled, zero code paths are affected.
// ============================================================================

use std::collections::HashMap;
use serde::{Deserialize, Serialize};

#[cfg(feature = "evolution")]
use rusqlite::Connection;

// ----------------------------------------------------------------------------
// Core structs
// ----------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeedbackSignal {
    pub memory_ids: Vec<i64>,
    pub edge_pairs: Vec<(i64, i64)>,
    pub useful: bool,
    pub timestamp: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvolutionState {
    pub node_weights: HashMap<i64, f32>,
    pub edge_weights: HashMap<(i64, i64), f32>,
    pub feedback_buffer: Vec<FeedbackSignal>,
    pub learning_rate: f32,
    pub generation: u32,
}

// Intermediate form for JSON persistence (HashMap keys must be strings in JSON)
#[derive(Debug, Serialize, Deserialize)]
struct EvolutionStateSerde {
    node_weights: HashMap<String, f32>,
    edge_weights: HashMap<String, f32>,
    learning_rate: f32,
    generation: u32,
}

impl EvolutionState {
    // Default weights 1.0, learning rate 0.01
    pub fn new() -> Self {
        EvolutionState {
            node_weights: HashMap::new(),
            edge_weights: HashMap::new(),
            feedback_buffer: Vec::new(),
            learning_rate: 0.01,
            generation: 0,
        }
    }

    // Add a feedback signal to the buffer
    pub fn record_feedback(&mut self, signal: FeedbackSignal) {
        self.feedback_buffer.push(signal);
    }

    // Process all buffered feedback signals.
    // Positive signal: weights += learning_rate, capped at 2.0
    // Negative signal: weights -= learning_rate, floored at 0.1
    pub fn train_step(&mut self) {
        let signals: Vec<FeedbackSignal> = self.feedback_buffer.drain(..).collect();

        for signal in &signals {
            let delta = if signal.useful {
                self.learning_rate
            } else {
                -self.learning_rate
            };

            for &id in &signal.memory_ids {
                let w = self.node_weights.entry(id).or_insert(1.0);
                *w = (*w + delta).clamp(0.1, 2.0);
            }

            for &pair in &signal.edge_pairs {
                let w = self.edge_weights.entry(pair).or_insert(1.0);
                *w = (*w + delta).clamp(0.1, 2.0);
            }
        }

        self.generation += 1;
    }

    // Returns learned node weight or 1.0 default
    pub fn get_node_weight(&self, id: i64) -> f32 {
        *self.node_weights.get(&id).unwrap_or(&1.0)
    }

    // Returns learned edge weight or 1.0 default
    pub fn get_edge_weight(&self, source: i64, target: i64) -> f32 {
        *self.edge_weights.get(&(source, target)).unwrap_or(&1.0)
    }

    // Serialize to the JSON-friendly form (keys as strings)
    fn to_serde(&self) -> EvolutionStateSerde {
        EvolutionStateSerde {
            node_weights: self.node_weights.iter()
                .map(|(k, v)| (k.to_string(), *v))
                .collect(),
            edge_weights: self.edge_weights.iter()
                .map(|((s, t), v)| (format!("{},{}", s, t), *v))
                .collect(),
            learning_rate: self.learning_rate,
            generation: self.generation,
        }
    }

    // Deserialize from the JSON-friendly form
    fn from_serde(s: EvolutionStateSerde) -> Self {
        let node_weights: HashMap<i64, f32> = s.node_weights.iter()
            .filter_map(|(k, v)| k.parse::<i64>().ok().map(|id| (id, *v)))
            .collect();

        let edge_weights: HashMap<(i64, i64), f32> = s.edge_weights.iter()
            .filter_map(|(k, v)| {
                let mut parts = k.splitn(2, ',');
                let a = parts.next()?.parse::<i64>().ok()?;
                let b = parts.next()?.parse::<i64>().ok()?;
                Some(((a, b), *v))
            })
            .collect();

        EvolutionState {
            node_weights,
            edge_weights,
            feedback_buffer: Vec::new(),
            learning_rate: s.learning_rate,
            generation: s.generation,
        }
    }

    // Persist to brain_meta table as JSON
    #[cfg(feature = "evolution")]
    pub fn save_state(&self, conn: &Connection) -> Result<(), String> {
        let serde_form = self.to_serde();
        let json_str = serde_json::to_string(&serde_form)
            .map_err(|e| format!("evolution serialize failed: {}", e))?;
        let blob = json_str.into_bytes();
        conn.execute(
            "INSERT OR REPLACE INTO brain_meta (key, value, updated_at) VALUES ('evolution_state', ?1, datetime('now'))",
            rusqlite::params![blob],
        ).map_err(|e| format!("evolution save failed: {}", e))?;
        Ok(())
    }

    // Load from brain_meta table. Returns fresh state if not found.
    #[cfg(feature = "evolution")]
    pub fn load_state(conn: &Connection) -> Self {
        use rusqlite::OptionalExtension;
        let result: rusqlite::Result<Option<Vec<u8>>> = conn.query_row(
            "SELECT value FROM brain_meta WHERE key = 'evolution_state'",
            [],
            |row| row.get(0),
        ).optional();

        match result {
            Ok(Some(blob)) => {
                match serde_json::from_slice::<EvolutionStateSerde>(&blob) {
                    Ok(s) => Self::from_serde(s),
                    Err(_) => Self::new(),
                }
            }
            _ => Self::new(),
        }
    }
}

impl Default for EvolutionState {
    fn default() -> Self {
        Self::new()
    }
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(all(test, feature = "evolution"))]
mod tests {
    use super::*;
    use tempfile::NamedTempFile;
    use rusqlite::Connection;

    fn open_test_db() -> (Connection, NamedTempFile) {
        let f = NamedTempFile::new().unwrap();
        let conn = Connection::open(f.path()).unwrap();
        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS brain_meta (
                key TEXT PRIMARY KEY,
                value BLOB,
                updated_at TEXT NOT NULL DEFAULT (datetime('now'))
            );"
        ).unwrap();
        (conn, f)
    }

    #[test]
    fn test_default_weights() {
        let state = EvolutionState::new();
        assert_eq!(state.get_node_weight(42), 1.0);
        assert_eq!(state.get_edge_weight(1, 2), 1.0);
        assert_eq!(state.learning_rate, 0.01);
        assert_eq!(state.generation, 0);
    }

    #[test]
    fn test_positive_feedback() {
        let mut state = EvolutionState::new();
        let signal = FeedbackSignal {
            memory_ids: vec![1, 2],
            edge_pairs: vec![(1, 2)],
            useful: true,
            timestamp: 0.0,
        };
        state.record_feedback(signal);
        state.train_step();

        let expected = 1.0_f32 + 0.01;
        assert!((state.get_node_weight(1) - expected).abs() < 1e-6,
            "node 1 weight expected ~{}, got {}", expected, state.get_node_weight(1));
        assert!((state.get_node_weight(2) - expected).abs() < 1e-6,
            "node 2 weight expected ~{}, got {}", expected, state.get_node_weight(2));
        assert!((state.get_edge_weight(1, 2) - expected).abs() < 1e-6,
            "edge (1,2) weight expected ~{}, got {}", expected, state.get_edge_weight(1, 2));
        assert_eq!(state.generation, 1);
    }

    #[test]
    fn test_negative_feedback() {
        let mut state = EvolutionState::new();
        let signal = FeedbackSignal {
            memory_ids: vec![5],
            edge_pairs: vec![(5, 6)],
            useful: false,
            timestamp: 0.0,
        };
        state.record_feedback(signal);
        state.train_step();

        let expected = 1.0_f32 - 0.01;
        assert!((state.get_node_weight(5) - expected).abs() < 1e-6,
            "node 5 weight expected ~{}, got {}", expected, state.get_node_weight(5));
        assert!((state.get_edge_weight(5, 6) - expected).abs() < 1e-6,
            "edge (5,6) weight expected ~{}, got {}", expected, state.get_edge_weight(5, 6));
    }

    #[test]
    fn test_weight_bounds() {
        let mut state = EvolutionState::new();
        // Drive a node weight to ceiling by 200 positive signals
        for _ in 0..200 {
            let signal = FeedbackSignal {
                memory_ids: vec![99],
                edge_pairs: vec![(99, 100)],
                useful: true,
                timestamp: 0.0,
            };
            state.record_feedback(signal);
            state.train_step();
        }
        assert!(state.get_node_weight(99) <= 2.0,
            "node weight must not exceed 2.0, got {}", state.get_node_weight(99));
        assert!((state.get_node_weight(99) - 2.0).abs() < 1e-4,
            "node weight should be at ceiling 2.0, got {}", state.get_node_weight(99));
        assert!(state.get_edge_weight(99, 100) <= 2.0,
            "edge weight must not exceed 2.0, got {}", state.get_edge_weight(99, 100));

        // Drive a node weight to floor by 200 negative signals
        let mut state2 = EvolutionState::new();
        for _ in 0..200 {
            let signal = FeedbackSignal {
                memory_ids: vec![77],
                edge_pairs: vec![(77, 78)],
                useful: false,
                timestamp: 0.0,
            };
            state2.record_feedback(signal);
            state2.train_step();
        }
        assert!(state2.get_node_weight(77) >= 0.1,
            "node weight must not go below 0.1, got {}", state2.get_node_weight(77));
        assert!((state2.get_node_weight(77) - 0.1).abs() < 1e-4,
            "node weight should be at floor 0.1, got {}", state2.get_node_weight(77));
    }

    #[test]
    fn test_save_load_roundtrip() {
        let (conn, _f) = open_test_db();

        let mut state = EvolutionState::new();
        state.node_weights.insert(10, 1.5);
        state.node_weights.insert(20, 0.7);
        state.edge_weights.insert((10, 20), 1.3);
        state.generation = 7;

        state.save_state(&conn).expect("save failed");

        let loaded = EvolutionState::load_state(&conn);
        assert_eq!(loaded.generation, 7);
        assert!((loaded.get_node_weight(10) - 1.5).abs() < 1e-6,
            "node 10 mismatch: got {}", loaded.get_node_weight(10));
        assert!((loaded.get_node_weight(20) - 0.7).abs() < 1e-6,
            "node 20 mismatch: got {}", loaded.get_node_weight(20));
        assert!((loaded.get_edge_weight(10, 20) - 1.3).abs() < 1e-6,
            "edge (10,20) mismatch: got {}", loaded.get_edge_weight(10, 20));
        // Unknown weights default to 1.0
        assert_eq!(loaded.get_node_weight(999), 1.0);
    }
}
